apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilderPartial.cs

```diff
@@ -57,14 +113,67 @@ public partial class SpannerClientBuilder
 
+        /// <summary>
+        /// Specifies the configuration to use for the affinity channel pool.
+        /// When set, each session will be bound to a gRPC channel meaning that
+        /// all commands executed on that session will be executed on that gRPC channel.
+        /// May be null, in which case there will be no channel affinity.
+        /// Defaults to null.
+        /// </summary>
+        public ChannelPoolConfig AffinityChannelPoolConfiguration {  get; set; }
+
+        /// <inheritdoc/>
+        /// <remarks>The default channel pool cannot be used when <see cref="AffinityChannelPoolConfiguration"/> is set.</remarks>
+        protected override bool CanUseChannelPool => base.CanUseChannelPool && AffinityChannelPoolConfiguration is null;
+
         private const string s_emulatorHostEnvironmentVariable = "SPANNER_EMULATOR_HOST";
         private static readonly string[] s_emulatorEnvironmentVariables = { s_emulatorHostEnvironmentVariable };
 
+        /// <inheritdoc/>>
+        /// <remarks>
+        /// Setting <see cref="AffinityChannelPoolConfiguration"/> to true is not compatible with specifying a <see cref="CallInvoker"/>.
+        /// </remarks>
+        protected override void Validate()
+        {
+            base.Validate();
+            GaxPreconditions.CheckState(CallInvoker is null || AffinityChannelPoolConfiguration is null, "Channel affinity cannot be configured with a custom CallInvoker.");
+        }
+
         partial void InterceptBuild(ref SpannerClient client) => client = MaybeCreateEmulatorClientBuilder()?.Build();
 
         partial void InterceptBuildAsync(CancellationToken cancellationToken, ref Task<SpannerClient> task) =>
             task = MaybeCreateEmulatorClientBuilder()?.BuildAsync(cancellationToken);
 
+        /// <inheritdoc/>
+        protected override CallInvoker CreateCallInvoker() =>
+         AffinityChannelPoolConfiguration is null
+            ? base.CreateCallInvoker()
+            : new GcpCallInvoker(
+                ServiceMetadata,
+                EffectiveEndpoint,
+                GetChannelCredentials(),
+                GetChannelOptions(),
+                GetApiConfing(),
+                EffectiveGrpcAdapter);
+
+        /// <inheritdoc/>
+        protected override async Task<CallInvoker> CreateCallInvokerAsync(CancellationToken cancellationToken) =>
+            AffinityChannelPoolConfiguration is null
+                ? await base.CreateCallInvokerAsync(cancellationToken).ConfigureAwait(false)
+                : new GcpCallInvoker(
+                    ServiceMetadata,
+                    EffectiveEndpoint,
+                    await GetChannelCredentialsAsync(cancellationToken).ConfigureAwait(false),
+                    GetChannelOptions(),
+                    GetApiConfing(),
+                    EffectiveGrpcAdapter);
+
+        private ApiConfig GetApiConfing() => new ApiConfig
+        {
+            ChannelPool = AffinityChannelPoolConfiguration.Clone(),
```

Comment:
I don't think we really need to clone here, if we're only passing this to the GcpCallInvoker constructor - that clones the ApiConfig anyway.